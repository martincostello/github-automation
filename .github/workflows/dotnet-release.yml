name: dotnet-release

on:
  schedule:
    - cron: "*/15 08-22 * * MON-FRI"
  workflow_dispatch:
    inputs:
      ref:
        description: "The dotnet/core ref to check for a new release against."
        required: false
        type: string
        default: ""

permissions: {}

jobs:
  check-for-release:
    name: check-for-release

    outputs:
      dotnet-core-updated-sha: ${{ steps.check-for-release.outputs.dotnet-core-updated-sha }}
      dotnet-releases-updated: ${{ steps.check-for-release.outputs.dotnet-releases-updated }}

    runs-on: [ ubuntu-latest ]

    steps:

      - name: "Get latest dotnet/core SHA and check for release notes changes"
        uses: actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410 # v6.4.1
        id: check-for-release
        with:
          script: |
            const owner = 'dotnet';
            const repo = 'core';

            const branch = await github.rest.repos.getBranch({
              owner,
              repo,
              branch: 'main',
            });

            const currentSha = branch.data.commit.sha;
            const previousSha = '${{ inputs.ref || vars.DOTNET_CORE_SHA }}';

            let releaseNotesFiles = [];
            let updatedSha = '';

            if (currentSha !== previousSha) {
              const diff = await github.rest.repos.compareCommitsWithBasehead({
                owner,
                repo,
                basehead: `${previousSha}...${currentSha}`,
              });
              updatedSha = currentSha;
              releaseNotesFiles = diff.data.files
                .map(file => file.filename)
                .filter(file => file.startsWith('release-notes/') && file.endsWith('/releases.json'));
            }

            console.log(JSON.stringify(releaseNotesFiles));

            const releaseNotesUpdated = releaseNotesFiles.length > 0;
            const releaseNotes = [];

            if (releaseNotesUpdated) {
              for (let path in releaseNotesFiles) {
                const content = await github.rest.repos.getContent({
                  owner,
                  repo,
                  path,
                  ref: currentSha,
                });
                releaseNotes.push(
                  JSON.parse(
                    Buffer.from(content.data.content, 'base64').toString()
                  )
                );
              }
            }

            core.setOutput('dotnet-core-updated-sha', updatedSha);
            core.setOutput('dotnet-release-notes', JSON.stringify(releaseNotes));
            core.setOutput('dotnet-releases-updated', releaseNotesUpdated);

  update-sha:
    name: update-sha
    needs: check-for-release
    if: ${{ needs.check-for-release.outputs.dotnet-core-updated-sha != '' }}

    concurrency:
      group: "update-sha"
      cancel-in-progress: false

    runs-on: [ ubuntu-latest ]

    steps:
      - name: "Update DOTNET_CORE_SHA"
      #  uses: actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410 # v6.4.1
      #  with:
      #    github-token: ${{ secrets.ACCESS_TOKEN }}
      #    script: |
      #      await github.rest.actions.updateRepoVariable({
      #        owner: context.repo.owner,
      #        repo: context.repo.repo,
      #        name: 'DOTNET_CORE_SHA',
      #        value: '${{ needs.check-for-release.outputs.dotnet-core-updated-sha }}'
      #      });
        shell: pwsh
        env:
          GH_TOKEN: ${{ secrets.ACCESS_TOKEN }}
        run: |
          gh api --method PATCH "/repos/${{ github.repository }}/actions/variables/DOTNET_CORE_SHA" -f value='${{ needs.check-for-release.outputs.dotnet-core-updated-sha }}'

  create-dispatch:
    name: create-dispatch
    needs: check-for-release
    if: ${{ needs.check-for-release.outputs.dotnet-releases-updated == 'true' }}

    permissions:
      contents: write

    runs-on: [ ubuntu-latest ]

    steps:
      - name: "Create repository dispatch for dotnet_release"
        uses: actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410 # v6.4.1
        with:
          script: |
            const releaseNotes = JSON.parse(`${{ needs.check-for-release.outputs.dotnet-release-notes }}`);
            const isPreview = (release) => {
              switch (release['support-phase']) {
                case 'go-live':
                case 'preview':
                  return true;
                default:
                  return false;
              }
            };

            const branchesToDispatch = [];
            if (releaseNotes.some(release => isPreview(release)) {
              branchesToDispatch.push('dotnet-vnext');
            }
            if (releaseNotes.some(release => !isPreview(release)) {
              branchesToDispatch.push('main');
            }

            for (let branch in branchesToDispatch) {
              await github.rest.repos.createDispatchEvent({
                owner: context.repo.owner,
                repo: context.repo.repo,
                event_type: 'dotnet_release',
                client_payload: {
                  branch
                }
              });
            }
