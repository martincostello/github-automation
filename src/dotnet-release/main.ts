// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { getFileContents } from '../shared/github';

/* eslint-disable no-console */

export async function run(): Promise<void> {
  try {
    const token = core.getInput('github-token', { required: false });
    const previousSha = core.getInput('ref', { required: false });
    const github = getOctokit(token);

    const owner = 'dotnet';
    const repo = 'core';

    const { data: branch } = await github.rest.repos.getBranch({
      owner,
      repo,
      branch: 'main',
    });

    const currentSha = branch.commit.sha;

    console.log(` Current SHA: ${currentSha}`);
    console.log(`Previous SHA: ${previousSha}`);

    let releaseNotesFiles: string[] = [];
    let updatedSha = '';

    if (currentSha !== previousSha) {
      const { data: diff } = await github.rest.repos.compareCommitsWithBasehead({
        owner,
        repo,
        basehead: `${previousSha}...${currentSha}`,
      });
      updatedSha = currentSha;
      if (diff.files) {
        releaseNotesFiles = diff.files
          .map((file) => file.filename)
          .filter((file) => file.startsWith('release-notes/') && file.endsWith('/releases.json'));
      }

      console.log(`${releaseNotesFiles} release notes file(s) were updated.`);
    }

    const releaseNotesUpdated = releaseNotesFiles.length > 0;
    const releaseNotes: any[] = [];

    if (releaseNotesUpdated) {
      for (const path of releaseNotesFiles) {
        const release = await getFileContents(github, owner, repo, path, currentSha);

        // Remove the releases array from the release notes
        // otherwise the JSON is too large to use as an output.
        const notes: ReleaseChannel = JSON.parse(release);
        delete notes.releases;

        console.log(`Release notes for ${path}:\n${JSON.stringify(notes, null, 2)}`);

        releaseNotes.push(notes);
      }
    }

    core.setOutput('dotnet-core-updated-sha', updatedSha);
    core.setOutput('dotnet-release-notes', JSON.stringify(releaseNotes));
    core.setOutput('dotnet-releases-updated', releaseNotesUpdated);
  } catch (error: any) {
    core.error(error);
    if (error instanceof Error) {
      core.setFailed(error.message);
    }
  }
}

if (require.main === module) {
  run();
}

interface ReleaseChannel {
  'channel-version': string;
  'latest-release': string;
  'latest-release-date': string;
  'latest-runtime': string;
  'latest-sdk': string;
  'release-type': string;
  'support-phase': string;
  'eol-date"': string;
  'lifecycle-policy"': string;
  'releases'?: Release[];
}

interface Release {
  'release-date': string;
  'release-version': string;
  'security': boolean;
  'release-notes': string;
}
