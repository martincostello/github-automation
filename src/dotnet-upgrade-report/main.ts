// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { fetch } from 'undici';
import { getBadge } from '../shared/badges';
import { ReleasesIndex } from '../shared/dotnet';
import { handle } from '../shared/errors';
import { getDotNetSdk, getFileContents, getPull, getReposForCurrentUser, getWorkflowConfig } from '../shared/github';

export async function run(): Promise<void> {
  try {
    const dotnetNextBranch = 'dotnet-vnext';

    const branch = (core.getInput('branch', { required: false }) || dotnetNextBranch).trim();
    const gist_id = core.getInput('gist-id', { required: false });
    const token = core.getInput('github-token', { required: false });
    const channel = core.getInput('channel', { required: false });

    const context = new Context();
    const github = getOctokit(token);

    const repositories = (await getReposForCurrentUser(github, 'all')).map((repo) => repo.full_name);
    const { checksOfInterest } = await getWorkflowConfig(github, context);

    let latestVersion = '';

    if (branch === dotnetNextBranch || !channel) {
      const releasesIndex = await getFileContents(github, 'dotnet', 'core', 'release-notes/releases-index.json', 'main');
      const releases: ReleasesIndex = JSON.parse(releasesIndex);
      if (releases['releases-index']?.length < 0) {
        throw new Error('No releases found in releases-index.json.');
      }
      latestVersion = releases['releases-index'][0]['latest-sdk'];
    } else {
      const quality = core.getInput('quality', { required: false }) || 'daily';
      const versionsFile = await fetch(`https://aka.ms/dotnet/${channel}/${quality}/productCommit-win-x64.json`);
      const versions = (await versionsFile.json()) as SdkProductCommits;
      if (versions?.sdk?.version) {
        latestVersion = versions.sdk.version.trim();
      }
    }

    const report = ['# .NET vNext Upgrade Report', ''];

    let count = 0;
    let wroteHeader = false;

    for (const slug of repositories) {
      // eslint-disable-next-line no-console
      console.log(`Fetching data for ${slug}.`);

      const [owner, repo] = slug.split('/');
      let commit_sha;
      try {
        const response = await github.rest.repos.getBranch({
          owner,
          repo,
          branch,
        });
        commit_sha = response.data.commit.sha;
      } catch {
        core.debug(`The ${branch} branch of ${slug} does not exist or does not have an open pull request.`);
        continue;
      }

      const dotnetSdk = await getDotNetSdk(github, owner, repo, branch);

      if (!dotnetSdk) {
        core.debug(`The ${branch} branch of ${slug} does not have a global.json file.`);
        continue;
      }

      const {
        data: { default_branch },
      } = await github.rest.repos.get({
        owner,
        repo,
      });

      const baseRefs = [default_branch];

      if (branch !== dotnetNextBranch) {
        baseRefs.push(dotnetNextBranch);
      }

      const { data: prs } = await github.rest.repos.listPullRequestsAssociatedWithCommit({
        owner,
        repo,
        commit_sha,
      });

      if (prs.length < 1) {
        continue;
      }

      const pull_for_ref = prs.find((pull) => baseRefs.includes(pull.base.ref));
      if (!pull_for_ref) {
        core.debug(
          `The ${branch} branch of ${slug} does not have an open pull request targeting any of the following branch(es): ${baseRefs.join(
            ','
          )}.`
        );
        continue;
      }

      const pull = await getPull(github, owner, repo, pull_for_ref.number);
      const hasConflicts = pull.mergeable === false;

      const { data: checkStatuses } = await github.rest.checks.listForRef({
        owner,
        repo,
        ref: commit_sha,
      });

      const statusesOfInterest = checkStatuses.check_runs
        .filter((check) => checksOfInterest.includes(check.name))
        .map((check) => ({
          status: check.status,
          conclusion: check.conclusion,
        }));

      let combinedStatus = 'success';

      for (const status of statusesOfInterest) {
        if (status.conclusion === 'failure') {
          combinedStatus = 'failure';
          break;
        } else if (status.status !== 'completed') {
          combinedStatus = 'pending';
        }
      }

      const sdkVersion = dotnetSdk.version;

      const buildColor = combinedStatus === 'success' ? 'brightgreen' : combinedStatus === 'pending' ? 'yellow' : 'red';
      const buildBadge = getBadge('build', combinedStatus, buildColor, 'github');
      const buildUrl = pull.html_url;

      const purple = '512BD4';
      const sdkColor = sdkVersion === latestVersion ? purple : 'yellow';
      const sdkBadge = getBadge('SDK', sdkVersion, sdkColor, 'dotnet');
      const sdkUrl = `${context.serverUrl}/${slug}/blob/${branch}/global.json#L${dotnetSdk.line}`;

      const conflictsColor = hasConflicts ? 'red' : 'brightgreen';
      const conflictsBadge = getBadge('Conflicts', hasConflicts ? 'Yes' : 'No', conflictsColor, 'git');
      const conflictsUrl = pull.html_url;

      count++;

      if (!wroteHeader) {
        report.push('| Pull Request | SDK Version | Build Status | Conflicts? |');
        report.push('| :----------- | :---------- | :----------- | :--------: |');
        wroteHeader = true;
      }

      report.push(
        `| [${slug}#${pull.number}](${pull.html_url}) | [![.NET SDK version](${sdkBadge})](${sdkUrl}) | [![Build: ${combinedStatus}](${buildBadge})](${buildUrl}) | [![Merge conflicts?](${conflictsBadge})](${conflictsUrl}) |`
      );
    }

    if (count === 0) {
      report.push(`No pull requests found for the \`${branch}\` branch.`);
    }

    report.push('');
    report.push(
      `Generated by [GitHub Actions](${context.serverUrl}/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId})`
    );

    const reportText = report.join('\n');

    await core.summary.addRaw(reportText).write();

    if (gist_id) {
      const { data: gist } = await github.rest.gists.update({
        gist_id,
        files: {
          'summary.md': {
            content: reportText,
          },
        },
      });
      if (gist.html_url) {
        core.notice(gist.html_url);
      }
    }
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}

interface ProductCommit {
  commit: string;
  version: string;
}

interface SdkProductCommits {
  runtime: ProductCommit;
  aspnetcore: ProductCommit;
  windowsdesktop: ProductCommit;
  sdk: ProductCommit;
}
