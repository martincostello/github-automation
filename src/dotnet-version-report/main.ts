// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { getBadge } from '../shared/badges';
import { handle } from '../shared/errors';
import { getFileContents, getDotNetSdk, getReposForCurrentUser } from '../shared/github';
import { ReleaseChannel } from '../shared/dotnet';

/* eslint-disable no-console */

export async function run(): Promise<void> {
  try {
    const token = core.getInput('github-token', { required: false });
    const context = new Context();
    const github = getOctokit(token);

    const repos = await getReposForCurrentUser(github, 'owner');

    const releases: ReleaseChannel[] = JSON.parse(
      await getFileContents(github, 'dotnet', 'core', 'release-notes/releases-index.json', 'main')
    );

    const report = ['# .NET SDK Version Report', '', '| Repository | SDK Version |', '| :--------- | :---------- |'];

    for (const repository of repos) {
      const slug = repository.full_name;
      const branch = repository.default_branch;

      console.log(`Fetching data for ${slug}.`);

      const dotnetSdk = await getDotNetSdk(github, repository.owner, repository.repo, branch);

      if (!dotnetSdk) {
        console.log(`No global.json found in ${slug}.`);
        continue;
      }

      const sdkVersion = dotnetSdk.version;
      const parts = sdkVersion.split('.');
      const channel = `${parts[0]}.${parts[1]}`;

      const [latestVersion] = releases['releases-index'].filter((p) => p['channel-version'] === channel).map((p) => p['latest-sdk']);

      if (!latestVersion) {
        console.info(`No latest version found for channel ${channel}`);
        continue;
      }

      const purple = '512BD4';
      const sdkColor = sdkVersion === latestVersion || latestVersion.startsWith(sdkVersion) ? purple : 'yellow';
      const sdkBadge = getBadge('SDK', sdkVersion, sdkColor, 'dotnet');
      const sdkUrl = `${context.serverUrl}/${slug}/blob/${branch}/global.json#L${dotnetSdk.line}`;

      report.push(`| [${slug}](${repository.html_url}) | [![.NET SDK version](${sdkBadge})](${sdkUrl}) |`);
    }

    await core.summary.addRaw(report.join('\n')).write();
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
