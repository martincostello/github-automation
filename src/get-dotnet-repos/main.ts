// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { handle } from '../shared/errors';
import { getDotNetSdk, getReposForCurrentUser } from '../shared/github';

type DotNetRepository = {
  ref: string;
  repo: string;
};

export async function run(): Promise<void> {
  try {
    const branch = core.getInput('branch', { required: false });
    const token = core.getInput('github-token', { required: false });
    const github = getOctokit(token);

    const ref = branch;
    const repositories = await getReposForCurrentUser(github, 'member');

    let singleRepository = core.getInput('repository', { required: false });

    if (singleRepository && !singleRepository.includes('/')) {
      const context = new Context();
      singleRepository = `${context.repo.owner}/${singleRepository}`;
    }

    const result: DotNetRepository[] = [];

    for (const repository of repositories) {
      const full_name = repository.full_name;

      if (singleRepository && singleRepository !== full_name) {
        core.debug(`Skipping repository ${full_name}.`);
        continue;
      }

      const owner = repository.owner;
      const repo = repository.repo;

      try {
        // eslint-disable-next-line no-console
        console.log(`Fetching data for ${full_name}.`);

        if (!(await getDotNetSdk(github, owner, repo, branch))) {
          core.debug(`The ${branch} branch of ${full_name} does not exist or does not have a global.json file.`);
          continue;
        }

        const config: DotNetRepository = {
          ref,
          repo: full_name,
        };

        result.push(config);
      } catch (err) {
        core.debug(`Failed to get data for ${full_name}: ${err}`);
      }
    }

    core.setOutput('repositories', JSON.stringify(result));
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
