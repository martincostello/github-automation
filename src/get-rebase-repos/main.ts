// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { getPull, getReposForCurrentUser } from '../shared/github';

export async function run(): Promise<void> {
  try {
    const headBranch = core.getInput('branch', { required: true });
    const baseBranch = core.getInput('base', { required: false });
    const force = core.getInput('force', { required: false }) === 'true';
    const specificRepo = core.getInput('repository', { required: false }) || '';
    const token = core.getInput('github-token', { required: false });

    const github = getOctokit(token);

    let repositories: string[];

    if (specificRepo) {
      repositories = [specificRepo];
    } else {
      repositories = (await getReposForCurrentUser({ octokit: github }, 'owner')).map((repo) => repo.full_name);
    }

    const result: string[] = [];
    for (const slug of repositories) {
      // eslint-disable-next-line no-console
      console.log(`Fetching data for ${slug}.`);

      const [owner, repo] = slug.split('/');

      let commit_sha;
      try {
        const { data: branch } = await github.rest.repos.getBranch({
          owner,
          repo,
          branch: headBranch,
        });
        commit_sha = branch.commit.sha;
      } catch (err) {
        core.debug(`Could not get branch ${headBranch}: ${err}`);
        continue;
      }

      const { data: prs } = await github.rest.repos.listPullRequestsAssociatedWithCommit({
        owner,
        repo,
        commit_sha,
      });

      if (prs.length < 1) {
        continue;
      }

      let baseRef = baseBranch;
      if (baseRef === '') {
        const { data: repository } = await github.rest.repos.get({
          owner,
          repo,
        });
        baseRef = repository.default_branch;
      }

      let rebaseBranch = force;
      if (!rebaseBranch) {
        const pull_for_ref = prs.find((pull) => pull.base.ref === baseRef);
        if (!pull_for_ref) {
          core.debug(`No pull request found targeting ${baseRef}.`);
          continue;
        }

        rebaseBranch = (await getPull(github, owner, repo, pull_for_ref.number)).mergeable_state === 'dirty';
        if (rebaseBranch) {
          core.notice(`${slug}:${headBranch} needs rebasing.`);
        }
      }

      if (rebaseBranch) {
        result.push(slug);
      }
    }

    core.setOutput('repositories', JSON.stringify(result));
  } catch (error: any) {
    core.error(error);
    if (error instanceof Error) {
      if (error.stack) {
        core.error(error.stack);
      }
      core.setFailed(error.message);
    }
  }
}

if (require.main === module) {
  run();
}
