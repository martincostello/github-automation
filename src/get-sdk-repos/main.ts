// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { UpdateDotNetSdkConfig } from '../shared/config';
import { handle } from '../shared/errors';
import { getDotNetSdk, getReposForCurrentUser, getUpdateConfiguration } from '../shared/github';

type UpdateConfiguration = {
  'channel': string;
  'exclude-nuget-packages': string | undefined;
  'include-nuget-packages': string | undefined;
  'labels': string;
  'quality': string;
  'ref': string;
  'repo': string;
  'update-nuget-packages': boolean;
};

function getConfiguration(overrides: UpdateDotNetSdkConfig | null): UpdateDotNetSdkConfig {
  const config: UpdateDotNetSdkConfig = {};

  if (overrides) {
    core.debug(`Baseline configuration: ${JSON.stringify(config)}`);
    core.debug(`Repository configuration: ${JSON.stringify(overrides)}`);

    const keys = ['channel', 'exclude-nuget-packages', 'include-nuget-packages', 'labels', 'quality', 'update-nuget-packages'];

    for (const key of keys) {
      const value = overrides[key];
      if (value !== undefined) {
        config[key] = value;
      }
    }

    core.debug(`Merged configuration: ${JSON.stringify(config)}`);
  }

  return config;
}

export async function run(): Promise<void> {
  try {
    const branch = core.getInput('branch', { required: false });
    const token = core.getInput('github-token', { required: false });
    const nightlyChannel = core.getInput('nightly-channel', { required: false });
    const github = getOctokit(token);

    const repositories = await getReposForCurrentUser(github, 'member');

    const excludePackages = '';
    const includePackages = 'Microsoft.AspNetCore.,Microsoft.EntityFrameworkCore.,Microsoft.Extensions.,System.Text.Json';
    const labels = 'dependencies,.NET';
    const ref = branch;
    const channel = branch === 'dotnet-nightly' && nightlyChannel ? nightlyChannel : '';
    const quality = branch === 'dotnet-nightly' ? 'daily' : '';

    let singleRepository = core.getInput('repository', { required: false });

    if (singleRepository && !singleRepository.includes('/')) {
      const context = new Context();
      singleRepository = `${context.repo.owner}/${singleRepository}`;
    }

    const result: UpdateConfiguration[] = [];

    for (const repository of repositories) {
      const full_name = repository.full_name;

      if (singleRepository && singleRepository !== full_name) {
        core.debug(`Skipping repository ${full_name}.`);
        continue;
      }

      const owner = repository.owner;
      const repo = repository.repo;

      const repoConfig = await getUpdateConfiguration(github, owner, repo, ref);
      if (repoConfig?.ignore) {
        core.debug(`Ignoring ${full_name}.`);
        continue;
      }

      try {
        // eslint-disable-next-line no-console
        console.log(`Fetching data for ${full_name}.`);

        if (!(await getDotNetSdk(github, owner, repo, branch))) {
          core.debug(`The ${branch} branch of ${full_name} does not exist or does not have a global.json file.`);
          continue;
        }

        const updateConfig = getConfiguration(repoConfig);

        const valueOrDefault = <T>(value: T | undefined, defaultValue: T): T => {
          return value !== undefined ? value : defaultValue;
        };

        const config: UpdateConfiguration = {
          channel,
          'exclude-nuget-packages': valueOrDefault(updateConfig['exclude-nuget-packages'], excludePackages),
          'include-nuget-packages': valueOrDefault(updateConfig['include-nuget-packages'], includePackages),
          labels,
          quality,
          ref,
          'repo': full_name,
          'update-nuget-packages': valueOrDefault(updateConfig['update-nuget-packages'], true),
        };

        result.push(config);
      } catch (err) {
        core.debug(`Failed to get data for ${full_name}: ${err}`);
      }
    }

    core.setOutput('updates', JSON.stringify(result));
  } catch (error: any) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
