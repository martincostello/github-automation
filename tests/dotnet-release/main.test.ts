// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import { ActionFixture } from '../ActionFixture';
import { run } from '../../src/dotnet-release/main';
import { setup } from '../fixtures';

describe('dotnet-release', () => {
  describe.each([
    ['no-new-commits', ''],
    ['no-new-commits', '6dc7506c307ddf3165cdd724ce1a5db4b31040a1'],
    ['no-releases', '53c167c2b0996ece03d56e20ff5977fdd16f3f9a'],
    ['stable', '02369e7047c776f8cbf06aab76a7e8a79064bef9'],
    ['preview', '02369e7047c776f8cbf06aab76a7e8a79064bef9'],
    ['multiple', '02369e7047c776f8cbf06aab76a7e8a79064bef9'],
  ])('%s for ref %s', (name: string, ref: string) => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      await setup(`dotnet-release/${name}`);
      await setup('dotnet-release/update-sha');

      fixture = new ActionFixture(run);
      await fixture.run({
        'github-token': 'fake-token',
        'state-token': 'actions-token',
        'ref': ref,
      });
    });

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });
  });
});
