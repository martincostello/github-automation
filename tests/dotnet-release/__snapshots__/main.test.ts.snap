// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dotnet-release multiple for ref 02369e7047c776f8cbf06aab76a7e8a79064bef9 outputs the correct SDK versions 1`] = `"[null,"7.0.400","8.0.100-preview.7.23376.3"]"`;

exports[`dotnet-release no-new-commits for ref  outputs the correct SDK versions 1`] = `"[]"`;

exports[`dotnet-release no-new-commits for ref 6dc7506c307ddf3165cdd724ce1a5db4b31040a1 outputs the correct SDK versions 1`] = `"[]"`;

exports[`dotnet-release no-releases for ref 53c167c2b0996ece03d56e20ff5977fdd16f3f9a outputs the correct SDK versions 1`] = `"[]"`;

exports[`dotnet-release preview for ref 02369e7047c776f8cbf06aab76a7e8a79064bef9 outputs the correct SDK versions 1`] = `"["8.0.100-preview.7.23376.3"]"`;

exports[`dotnet-release stable for ref 02369e7047c776f8cbf06aab76a7e8a79064bef9 outputs the correct SDK versions 1`] = `"[null]"`;
